src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs

```diff
@@ -2358,6 +2358,69 @@ var projectUniqueNamesForBuildIntToUpdate
                         }
                     }
 
+#if PACKAGE_CONFIG_LOCK
+                    try
+                    {
+                        if (msbuildProject != null)
+                        {
+                            //start batch processing for msbuild
+                            await msbuildProject.ProjectSystem.BeginProcessingAsync();
+                        }
+
+                        //todo: use same logic as PackageReference code to enable packages lock file (msbuild prop or file already exists)
+                        var useLockFile = false;
+                        var property = msbuildProject?.ProjectSystem?.GetPropertyValue("RestorePackagesWithLockFile");
+                        if (property is string propertyValue)
+                        {
+                            bool.TryParse(propertyValue, out useLockFile);
+                        }
+                        if (useLockFile)
+                        {
+                            //todo: get lock file name same way PackageReference projects do
+                            var directory = Path.GetDirectoryName(msbuildProject.MSBuildProjectPath);
+                            var lockFileFullPath = Path.Combine(directory, PackagesLockFileFormat.LockFileName);
+                            var lockFileName = Path.GetFileName(lockFileFullPath);
+
+                            var installedPackagesEnumeration = await nuGetProject.GetInstalledPackagesAsync(token);
+                            var installedPackages = installedPackagesEnumeration as ICollection<PackageReference> ?? installedPackagesEnumeration.ToList();
```

Comment:
why did you need to type cast it to `ICollection`?

Reply:
I was doing it because Linq's `.ToList()` is expensive, particularly for large lists. If object is an ICollection we can cast much more cheaply than creating a new list and possibly resizing multiple times.

Reply:
why do you even need to convert it to `List`? just pass the `IEnumerable` down the line...

Reply:
The original code needed to know if the result was empty or not. Multiple enumerations of IEnumerable throws exceptions (for example calling .Any() followed by a foreach), depending on data source, so I avoid doing anywhere as a rule.

However, I changed the code to keep the lock file when there are no packages installed. It creates a lock file without any targets or packages. I did this to ensure if that if someone opts-in to lock files, then uninstalls all packages and then installs a package, they're still opted-in to lock files. So, it no longer needs different code paths for zero packages, so this cast/check is no longer done.